ArcGISRuntimeSDKQt_CppSamples/Maps/HonorMobileMapPackageExpiration/HonorMobileMapPackageExpiration.cpp

```diff
@@ -0,0 +1,170 @@
+  QObject(parent),
+  m_dataPath(defaultDataPath() + sampleMmpk)
+{
+  // connect to the Mobile Map Package instance to determine if direct read is supported. Packages
+  // that contain raster data cannot be read directly and must be unpacked first.
+  connect(MobileMapPackage::instance(), &MobileMapPackage::isDirectReadSupportedCompleted,
+          this, [this](QUuid, bool supported)
+  {
+    // if direct read is supported, load the package
+    if (supported)
+    {
+      createMapPackage(m_dataPath);
+    }
+    // otherwise, the package needs to be unpacked
+    else
+    {
+      MobileMapPackage::unpack(m_dataPath, m_unpackTempDir.path());
+    }
+  });
+
+  // connect to the Mobile Map Package instance to know when the data is unpacked
+  connect(MobileMapPackage::instance(), &MobileMapPackage::unpackCompleted,
+          this, [this](QUuid, bool success)
+  {
+    // if the unpack was successful, load the unpacked package
+    if (success)
+    {
+      createMapPackage(m_unpackTempDir.path());
+    }
+    // log that the upack failed
+    else
+    {
+      qDebug() << "failed to unpack";
+    }
+  });
+
+  // connect to the Mobile Map Package instance to know when errors occur
+  connect(MobileMapPackage::instance(), &MobileMapPackage::errorOccurred,
+          [](Error e)
+  {
+    if (e.isEmpty())
+    {
+      return;
+    }
+
+    qDebug() << QString("Error: %1 %2").arg(e.message(), e.additionalMessage());
+  });
+
+}
+
+HonorMobileMapPackageExpiration::~HonorMobileMapPackageExpiration() = default;
+
+void HonorMobileMapPackageExpiration::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<HonorMobileMapPackageExpiration>("Esri.Samples", 1, 0, "HonorMobileMapPackageExpirationSample");
+}
+
+MapQuickView* HonorMobileMapPackageExpiration::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void HonorMobileMapPackageExpiration::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+
+  emit mapViewChanged();
+
+  // Check if the MMPK can be read directly
+  MobileMapPackage::isDirectReadSupported(m_dataPath);
+}
+
+void HonorMobileMapPackageExpiration::createMapPackage(const QString& path)
+{
+  // instatiate a mobile map package
+  m_mobileMapPackage = new MobileMapPackage(path, this);
+
+  // wait for the mobile map package to load
+  connect(m_mobileMapPackage, &MobileMapPackage::doneLoading, this, [this](Error error)
+  {
+    if (!error.isEmpty())
+    {
+      qDebug() << QString("Package load error: %1 %2").arg(error.message(), error.additionalMessage());
+      return;
+    }
+
+    if (!m_mobileMapPackage || !m_mapView || m_mobileMapPackage->maps().isEmpty())
+    {
+      return;
+    }
+
+    // Access Expiration Info
+    Expiration expiration = m_mobileMapPackage->expiration();
```

Comment:
Should we test to see if the returned expiration is empty before proceeding to insure it is valid? 

if (expiration.isEmpty())
  // return and do nothing?

Reply:
good call. added that in to the latest commit